File: src/main/scala/gd/QueryEvaluatorPerformanceTest.scala
package gd

import java.util.concurrent.TimeUnit
import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra.Blackhole

@OutputTimeUnit(TimeUnit.MILLISECONDS)
@BenchmarkMode(Array(Mode.AverageTime))
class QueryEvaluatorPerformanceTest {

  @Benchmark
  def test(state: BenchmarkState, bh: Blackhole): Unit = {
    val results = state.evaluator.findAllAttributesMatch(Seq(Attributes.MALE, Attributes.SINGLE), Seq(Attributes.SINGLE))
    bh.consume(results)
  }
}

@State(Scope.Thread)
class BenchmarkState {
  private val encoder = new BitwiseEncoder
  private val testPersonBuilder = new TestPersonBuilder(encoder)
  def evaluator = {
    val input = 1 to 1000000 map testPersonBuilder.buildSamplePerson
    new QueryEvaluator(input, encoder)
  }
}

File: src/main/java/gd/QueryEvaluatorPerformanceTest.java
package gd;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;

@OutputTimeUnit(TimeUnit.MILLISECONDS)
@BenchmarkMode(Mode.AverageTime)
public class QueryEvaluatorPerformanceTest {

  @Benchmark
  public void test(BenchmarkState state, Blackhole bh) {
    List<Attributes> filterAttributes = new ArrayList<Attributes>() {{
      this.add(Attributes.MALE);
      this.add(Attributes.FEMALE);
    }};

    List<Attributes> groupByAttributes = new ArrayList<Attributes>() {{
      this.add(Attributes.SINGLE);
    }};
    GroupByRollupResult results = state.evaluator.findAllAttributesMatch(filterAttributes, groupByAttributes);
    bh.consume(results);
  }
}
